File: ios-app/AnagramArena/Sources/ViewModels/ConundrumPracticeViewModel.swift
import Combine
import Foundation

final class ConundrumPracticeViewModel: ObservableObject {
    enum Phase {
        case ready
        case solving
        case result
    }

    @Published private(set) var phase: Phase = .ready
    @Published private(set) var conundrum: Conundrum?
    @Published var guess: String = ""
    @Published private(set) var secondsRemaining: Int = 30
    @Published private(set) var result: ConundrumRoundResult?

    private let provider: ConundrumProviding
    private let validator: ConundrumValidator
    private let timerEnabled: Bool
    private let roundDuration: Int
    private var timerCancellable: AnyCancellable?

    init(
        provider: ConundrumProviding,
        timerEnabled: Bool,
        roundDuration: Int = 30,
        validator: ConundrumValidator = ConundrumValidator()
    ) {
        self.provider = provider
        self.timerEnabled = timerEnabled
        self.roundDuration = roundDuration
        self.validator = validator
        self.secondsRemaining = roundDuration

        startRound()
    }

    var canSubmit: Bool {
        phase == .solving && conundrum != nil
    }

    func startRound() {
        stopTimer()

        conundrum = provider.randomConundrum()
        guess = ""
        secondsRemaining = roundDuration
        result = nil
        phase = conundrum == nil ? .ready : .solving

        guard timerEnabled, phase == .solving else { return }

        timerCancellable = Timer.publish(every: 1, on: .main, in: .common)
            .autoconnect()
            .sink { [weak self] _ in
                self?.onTick()
            }
    }

    func submit() {
        guard phase == .solving, let conundrum else { return }

        stopTimer()
        let solved = validator.isCorrect(guess: guess, answer: conundrum.answer)
        result = ConundrumRoundResult(
            conundrum: conundrum,
            submittedGuess: guess,
            solved: solved,
            score: solved ? 12 : 0
        )
        phase = .result
    }

    private func onTick() {
        guard phase == .solving else {
            stopTimer()
            return
        }

        if secondsRemaining > 0 {
            secondsRemaining -= 1
        }

        if secondsRemaining == 0 {
            submit()
        }
    }

    private func stopTimer() {
        timerCancellable?.cancel()
        timerCancellable = nil
    }
}

File: android-app/app/src/main/java/com/bhan796/anagramarena/viewmodel/ConundrumPracticeViewModel.kt
package com.bhan796.anagramarena.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.bhan796.anagram.core.model.Conundrum
import com.bhan796.anagram.core.model.ConundrumRoundResult
import com.bhan796.anagram.core.validation.ConundrumValidator
import com.bhan796.anagramarena.data.ConundrumProvider
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

enum class ConundrumPracticePhase {
    READY,
    SOLVING,
    RESULT
}

data class ConundrumPracticeUiState(
    val phase: ConundrumPracticePhase = ConundrumPracticePhase.READY,
    val conundrum: Conundrum? = null,
    val guess: String = "",
    val secondsRemaining: Int = 30,
    val result: ConundrumRoundResult? = null
) {
    val canSubmit: Boolean = phase == ConundrumPracticePhase.SOLVING && conundrum != null
}

class ConundrumPracticeViewModel(
    private val provider: ConundrumProvider,
    private val timerEnabled: Boolean,
    private val roundDuration: Int = 30,
    private val validator: ConundrumValidator = ConundrumValidator()
) : ViewModel() {
    private var timerJob: Job? = null

    private val _state = MutableStateFlow(ConundrumPracticeUiState(secondsRemaining = roundDuration))
    val state: StateFlow<ConundrumPracticeUiState> = _state

    init {
        startRound()
    }

    fun updateGuess(value: String) {
        _state.update { it.copy(guess = value) }
    }

    fun startRound() {
        stopTimer()
        val conundrum = provider.randomConundrum()

        _state.value = ConundrumPracticeUiState(
            phase = if (conundrum == null) ConundrumPracticePhase.READY else ConundrumPracticePhase.SOLVING,
            conundrum = conundrum,
            guess = "",
            secondsRemaining = roundDuration,
            result = null
        )

        if (timerEnabled && conundrum != null) {
            startTimer()
        }
    }

    fun submit() {
        val current = _state.value
        val conundrum = current.conundrum ?: return
        if (current.phase != ConundrumPracticePhase.SOLVING) return

        stopTimer()
        val solved = validator.isCorrect(current.guess, conundrum.answer)
        val result = ConundrumRoundResult(
            conundrum = conundrum,
            submittedGuess = current.guess,
            solved = solved,
            score = if (solved) 12 else 0
        )

        _state.update {
            it.copy(phase = ConundrumPracticePhase.RESULT, result = result)
        }
    }

    private fun startTimer() {
        timerJob = viewModelScope.launch {
            while (true) {
                delay(1000)
                val current = _state.value
                if (current.phase != ConundrumPracticePhase.SOLVING) {
                    stopTimer()
                    break
                }

                val remaining = (current.secondsRemaining - 1).coerceAtLeast(0)
                _state.update { it.copy(secondsRemaining = remaining) }

                if (remaining == 0) {
                    submit()
                    break
                }
            }
        }
    }

    private fun stopTimer() {
        timerJob?.cancel()
        timerJob = null
    }

    companion object {
        fun factory(provider: ConundrumProvider, timerEnabled: Boolean): ViewModelProvider.Factory {
            return object : ViewModelProvider.Factory {
                @Suppress("UNCHECKED_CAST")
                override fun <T : ViewModel> create(modelClass: Class<T>): T {
                    return ConundrumPracticeViewModel(provider = provider, timerEnabled = timerEnabled) as T
                }
            }
        }
    }
}